### PR TITLE
Dynamic Bar mock data and test cases

### PR DESCRIPTION
**Pull request by mayank-nx**
Hi @nxanil @natabal 
Please verify the test i have created for dynamic bar graph.

Thanks

### CODE REVIEW

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            global.config = await getDataAndConfig('DynamicBarGraph');     
+    });
+
+    describe("positive-data-horizontal-number", ()=>{
+        
+        beforeAll( async () => {
+            global.componenthorizontal = mount(<BarGraph width={500} height={500} configuration={config.horizontal} data={config.data}> </BarGraph>);
+        });
+
+        it("number of barGraph is matched", ()=>{
```
Comment (anilkchauhan) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Label must be like "Total Bars"

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            const middleHtml = componenthorizontal.find('.graph-bars').html();
+            const $ = cheerio.load(middleHtml);
+            const noOfBars = $('.graph-bars').find('g').length;
+            const totalBar = global.componenthorizontal.root.props().configuration.data.otherOptions.limit + 1;
+            expect(noOfBars).toBe(totalBar)
+        });
+
+        it("check xAxis ticks", ()=>{
+            const xAxisTicks = CheckTicks(componenthorizontal,'.graph-container','.xAxis','g')
+            expect(xAxisTicks).toBeGreaterThan(0);
```
Comment (anilkchauhan) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Must be compared with equality over here.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        
+        beforeAll( async () => {
+            global.componenthorizontal = mount(<BarGraph width={500} height={500} configuration={config.horizontal} data={config.data}> </BarGraph>);
+        });
+
+        it("number of barGraph is matched", ()=>{
+            const middleHtml = componenthorizontal.find('.graph-bars').html();
+            const $ = cheerio.load(middleHtml);
+            const noOfBars = $('.graph-bars').find('g').length;
+            const totalBar = global.componenthorizontal.root.props().configuration.data.otherOptions.limit + 1;
```
Comment (anilkchauhan) [inline Graphs/DynamicBarGraph/barGraph.test.js]: This line looks very odd. Look for alternative.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        });
+
+        it("check xAxis ticks", ()=>{
+            const xAxisTicks = CheckTicks(componenthorizontal,'.graph-container','.xAxis','g')
+            expect(xAxisTicks).toBeGreaterThan(0);
+        });
+    
+        it("check yAxis ticks", ()=>{
+            const yAxisTicks = CheckTicks(componenthorizontal,'.graph-container','.yAxis','g')
+            expect(yAxisTicks).toBeGreaterThan(0);
```
Comment (anilkchauhan) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Use Equality operator.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        });
+        
+    });   
+
+    describe("vertical-percentage", ()=>{
+        beforeAll( async () => {
+            global.componentVertical = mount(<BarGraph width={500} height={500} configuration={config.vertical} data={config.data}> </BarGraph>);
```
Comment (anilkchauhan) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Remove global from all the areas and files and use ` this.verticalBar = ..... `

Comment (anilkchauhan) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Also please take care of the lin length. Please split it into multiple lines.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        });
+    
+        it("check yAxis ticks", ()=>{
+            const yAxisTicks = CheckTicks(componenthorizontal,'.graph-container','.yAxis','g')
+            expect(yAxisTicks).toBeGreaterThan(0);
+        });
+        
+    });   
+
+    describe("vertical-percentage", ()=>{
```
Comment (anilkchauhan) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Change the description with better name. Avoid using "-" in names.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        beforeAll( async () => {
+            global.componentVertical = mount(<BarGraph width={500} height={500} configuration={config.vertical} data={config.data}> </BarGraph>); 
+        });
+
+        it("number of barGraph is matched", ()=>{
```
Comment (anilkchauhan) [inline Graphs/DynamicBarGraph/barGraph.test.js]: "Number of Bars" will do the Job

File: Graphs/testHelper.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        const configFolder = __dirname + '/' + graphName + '/mock/configurations';
+        let allFiles = await readdirAsync(configFolder);
+        let config = [];
+        allFiles.forEach(file => {
+            config[file.split(".")[0]] = readJSON(configFolder + '/' + file);
+        });
+        config['data'] = readJSON(dataFolder + '/data.json');
+        resolve(config);
+    });
+}
```
[the remark this answers is not in the document]
Comment (anilkchauhan) [reply Graphs/testHelper.js]: Please make these changes for all the files.

Comment (anilkchauhan) [inline Graphs/testHelper.js]: @mayank-nx 
- Comparison like Bar Width and Height is missing.
- Comparison like positions of the g is missing, that will help us to know that Bar is plotted at exact location.
- Compare it for two random bars.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+import {CheckTicks,getDataAndConfig} from '../testHelper';
+import BarGraph from '.'; 
+const cheerio = require('cheerio')
+
+describe ("test barGrpah", ()=>{
+    beforeAll( async () => {
+            global.config = await getDataAndConfig('DynamicBarGraph');     
+    });
+
+    describe("positive-data-horizontal-number", ()=>{
```
[the remark this answers is not in the document]
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        
+        beforeAll( async () => {
+            global.componenthorizontal = mount(<BarGraph width={500} height={500} configuration={config.horizontal} data={config.data}> </BarGraph>);
```
[the remark this answers is not in the document]
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js (hunk 1 above)
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        });
+
+        it("number of barGraph is matched", ()=>{
+            const middleHtml = componenthorizontal.find('.graph-bars').html();
```
[the remark this answers is not in the document]
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: variable name has been changed.

File: Graphs/DynamicBarGraph/barGraph.test.js (hunk 3 above)
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            const $ = cheerio.load(middleHtml);
+            const noOfBars = $('.graph-bars').find('g').length;
+            const totalBar = global.componenthorizontal.root.props().configuration.data.otherOptions.limit + 1;
+            expect(noOfBars).toBe(totalBar)
+        });
+
+        it("check xAxis ticks", ()=>{
+            const xAxisTicks = CheckTicks(componenthorizontal,'.graph-container','.xAxis','g')
```
[the remark this answers is not in the document]
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js (hunk 2 above)
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js (hunk 4 above)
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js (hunk 6 above)
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: all the "-" symbol removed.

File: Graphs/DynamicBarGraph/barGraph.test.js (hunk 5 above)
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js (hunk 7 above)
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+            const xAxisTicks = CheckTicks(componentWithoutBrush,'.graph-container','.xAxis','g')
+            expect(xAxisTicks).toBeGreaterThan(0);
+        });
+    
+        it("check yAxis ticks", ()=>{
+            const yAxisTicks = CheckTicks(componentWithoutBrush,'.graph-container','.yAxis','g')
+            expect(yAxisTicks).toBeGreaterThan(0);
+        }); 
+    });
+});
```
[the remark this answers is not in the document]
Comment (mayank-nx) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Done

File: Graphs/testHelper.js (hunk 8 above)
Comment (mayank-nx) [reply Graphs/testHelper.js]: I have created a test case in which I have covered all the above points.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 14]
```diff
@@ -0,0 +1,136 @@
+import React from 'react';
+import {mount} from 'enzyme';
+import {CheckTicks,getDataAndConfig} from '../testHelper';
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Can we have proper spacing here { mount }?

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+import BarGraph from '.'; 
+const cheerio = require('cheerio')
+
+describe ("test barGrpah", ()=>{
+    beforeAll( async () => {
+            global.config = await getDataAndConfig('DynamicBarGraph');
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Please take care of indentations.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 16]
```diff
@@ -0,0 +1,613 @@
...
+const cheerio = require('cheerio');
+
+describe("Bar Graph", () => {
+    let config;
+    beforeAll(async () => {
+        config = await getDataAndConfig('DynamicBarGraph');
+    });
+
+    describe("Vertical Bar", () => {
+
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Too many unwanted new lines.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 17]
```diff
@@ -0,0 +1,613 @@
...
+                        width={500}
+                        height={500}
+                        configuration={config.verticalNumber}
+                        data={config.data}>
+                    </BarGraph>
+                );
+                setTimeout(() => { 
+                    verticalNumber.update(); 
+                    done(); 
+                }, 350);
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Please add a comment here, to let the user know why you are using it.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 18]
```diff
@@ -0,0 +1,613 @@
...
+                    </BarGraph>
+                );
+                setTimeout(() => { 
+                    verticalNumber.update(); 
+                    done(); 
+                }, 350); 
+
+            });
+           
+            it("Test svg", () => {
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Please use a proper name, and avoid using tests.

Comment (nxanil) [reply Graphs/DynamicBarGraph/barGraph.test.js]: Like SVG Dimensions, and all other places as well.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 19]
```diff
@@ -0,0 +1,613 @@
...
+            });
+
+            it("Total Bars", () => {
+                const graphBar = verticalNumber.find('.graph-bars').html();
+                const $ = cheerio.load(graphBar);
+                const noOfBars = $('.graph-bars').find('g').length;
+                expect(noOfBars).toBe(6);
+            });
+
+            it("Check xAxis ticks", () => {
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Must be like "xAxis ticks length" - Name must let you know, what are you trying to do. "Check" is bit general.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 20]
```diff
@@ -0,0 +1,613 @@
...
+                const xAxisTicks = checkTicks(verticalNumber, '.graph-container', '.xAxis', 'g')
+                expect(xAxisTicks).toBe(6);
+            });
+
+            it("Check yAxis ticks", () => {
+                const yAxisTicks = checkTicks(verticalNumber, '.graph-container', '.yAxis', 'g')
+                expect(yAxisTicks).toBe(6);
+            });
+
+            it("Checking Bars Height Width X Y", () => {
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Must be like "Bar Dimensions - Height, width and positions"

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 21]
```diff
@@ -0,0 +1,613 @@
...
+                const width = parseFloat(bar.attr('width'));
+                expect(x).toBeCloseTo(80.07);
+                expect(y).toBeCloseTo(284.54);
+                expect(height).toBeCloseTo(167.46);
+                expect(width).toBeCloseTo(65.51);
+            });
+
+        });
+
+        describe("Percentage", () => {
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Not sure, what "percentage" is.

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 22]
```diff
@@ -0,0 +1,613 @@
...
+                        data={config.data}>
+                    </BarGraph>
+                );
+                setTimeout(() => { 
+                    verticalPercentage.update(); 
+                    done(); 
+                }, 350); 
+            });
+
+            it("Test svg", () => {
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Avoid using test and checks

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 23]
```diff
@@ -0,0 +1,613 @@
...
+                const width = parseFloat(bar.attr('width'));
+                expect(x).toBeCloseTo(48.36);
+                expect(y).toBeCloseTo(60.87);
+                expect(height).toBeCloseTo(391.13);
+                expect(width).toBeCloseTo(39.56);
+            });
+
+        });
+
+        describe("WithoutBush", () => {
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: why camelcase - it could be "Graph with Brush"

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 24]
```diff
@@ -0,0 +1,613 @@
...
+                const xAxisTicks = checkTicks(verticalWithoutBrush, '.graph-container', '.xAxis', 'g')
+                expect(xAxisTicks).toBe(10);
+            });
+
+            it("Check yAxis ticks", () => {
+                const yAxisTicks = checkTicks(verticalWithoutBrush, '.graph-container', '.yAxis', 'g')
+                expect(yAxisTicks).toBe(5);
+            });
+
+            it("Checking Bars Height Width X Y", () => {
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Can we make these methods generic? I can see lots of repetitions.

File: Graphs/testHelper.js [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+    const graph = component.find(tag).html();
+    const cheerioData = cheerio.load(graph);
+    return cheerioData;
+}
+
+export const checkSvg = (component) => {
+    const $ = getHtml(component, 'svg');
+    const svgHeight = $('svg').attr('height');
+    const svgWidth = $('svg').attr('width');
+    if(svgHeight == "500" && svgWidth == "500"){
```
Comment (nxanil) [inline Graphs/testHelper.js]: @mayank-nx Please take care of indentation and spacings and you could simply use:
`return svgHeight == "500" && svgWidth == "500";`

File: Graphs/DynamicBarGraph/barGraph.test.js [hunk 26]
```diff
@@ -0,0 +1,550 @@
...
+            });
+
+            it("yAxis Ticks Length", () => {
+                const yAxisTicks = checkTicks(verticalNumber, '.graph-container', '.yAxis', 'g')
+                expect(yAxisTicks).toBe(6);
+            });
+
+            it("Bar Dimensions Height, Width and Positions", () => {
+                $ = getHtml(verticalNumber, '.graph-bars');
+                const bar = $('.graph-bars').children().first().next().find('rect');
```
Comment (nxanil) [inline Graphs/DynamicBarGraph/barGraph.test.js]: Could we do it for at least two bars? and create a helper function to return the dimension of the bar as an object and then directly compare that object with lodash? and all other areas. That will reduce down the coding.

**Comment by nxanil** [comment]
@mayank-nx Two important things:
 - You are not checking whether brush has rendered properly or not.
 - In Case of stack bar graph, please check all the bars in a stack and at least two stack bars in full.
 - Also, legend testing is missing.

**Comment by nxanil** [comment]
@natabal Would you like to have a look at it, or do I merge it?

**Comment by natabal** [comment]
@nxanil Please merge!

**Comment by natabal** [comment]
@nxanil Let me have a quick look!

**Pull request merged**